feat(AWS Lambda): Support `Ref` for `destinations` properties

ALOHACREPES345 opened the pull request:
<!-- ⚠️⚠️ Acknowledge ALL below remarks -->
<!-- ⚠️⚠️ PR will not be processed if it doesn't meet outlined criteria -->

<!-- ⚠️⚠️ Do not propose PR's without prior agreement on a solution in the corresponding issue -->
<!-- ⚠️⚠️ Only documentation updates and obvious bug fixes are welcome without it -->

<!--
⚠️⚠️ Ensure to follow code style guidelines
https://github.com/serverless/serverless/blob/main/CONTRIBUTING.md#code-style
-->

<!--
⚠️⚠️ Ensure to cover changes with tests written according to test guidelines
https://github.com/serverless/serverless/blob/main/test/README.md
--

<!-- ⚠️⚠️ Ensure that support for Node.js v12 is maintained. -->

<!--
⚠️⚠️ Ensure that the proposed change passes CI. Confirm that by running the following scripts:
• npm run prettier-check
• npm run lint
• npm test
-->

<!--
⚠️⚠️ If proposed change touches integration with AWS services, confirm integration tests pass:
https://github.com/serverless/serverless/blob/main/test/README.md#aws-integration-tests
-->

<!-- ⚠️⚠️ After your PR is submitted, review the final CI status and address eventual failure -->

<!-- ⚠️⚠️ Answer below questions -->

<!--
Q1: Provide a link to the corresponding issue

• If PR *partially* addresses issue, ensure to rename "Closes" to "Addresses" ("Closes" term will automatically close an issue on PR merge)
• If it's a documentation update or obvious bug fix that has no corresponding issue, replace this line with a short description of made changes
-->

Closes: #10799

pgrzesik commented on lib/plugins/aws/package/compile/functions.js:
```diff
@@ -615,9 +618,12 @@ class AwsCompileFunctions {
-            ? destinations.onFailure
-            : this.provider.resolveFunctionArn(destinations.onFailure),
+          Destination:
+            (typeof destinations.onFailure === 'string' &&
```
this resolution logic sounds like a repetition between `onSuccess` and `onFailure` - maybe it could be abstracted to a small function that can be reused?

ALOHACREPES345 replied:
You're right. I'll fix it.

pgrzesik commented on lib/plugins/aws/package/compile/functions.js:
```diff
@@ -653,30 +659,42 @@ class AwsCompileFunctions {
+        if (functionAddress.includes(':event-bus/')) return 'events:PutEvents';
+      }
+      if (typeof functionAddress === 'object' && 'Ref' in functionAddress) {
+        const type = this.serverless.service.resources.Resources[functionAddress.Ref].Type;
```
This would first need to be first validated that `Ref` is actually referencing a resource that is defined in CloudFormation and I'm afraid that at this point the resources might not yet be available fully to be checked. Could you please validate that with a test and introduce a check that throws a meaningful error? Maybe the better way would be to explicitly require users to provide a type in such case? Trying to resolve such values from references is always tricky

pgrzesik replied:
Thanks for the proposal @ALOHACREPES345 - I think it would be easier to configure with such approach:

```
onSuccess:
  type: ...
  arn: ...
```

while at the same time, we would still support this syntax:

```
onSuccess: ...
```

but throw a descriptive error if someone tries to pass it as Ref for example. 

We already do something similar e.g. here: https://github.com/serverless/serverless/blob/6a6417c29fe6c06286562ab7604536004bf62334/lib/plugins/aws/package/compile/events/stream.js#L167-L184 where the type can be passed explicitly or it can be derived from passed configuration - in this case we don't throw errors, but I think in case of `destinations` we should keep things simple and just require `type` explicitly if someone passes an ARN

ALOHACREPES345 replied:
Thank you very much for sharing an example. 
I'll take a closer look at the code.

pgrzesik reviewed (changes requested):
thanks @ALOHACREPES345 - it looks like a good start, but I'm worried it might actually not work as intended in some cases, especially when trying to resolve proper type from the reference - could you please look into my comments and let me know what do you think?

ALOHACREPES345 commented:
Thank you for your help. @pgrzesik 
I fixed the code with following the existing code and based on your advice.
- added type property like stream.js
- added error message like sns.js and validate.js and the advice of `I think in case of destinations we should keep things simple and just require type explicitly if someone passes an ARN`
- fixed provider.js while referencing stream.js

But there is one thing I'd like to ask.
I try to add a test case with newly pushed commit.

```
    it('should throw error when `functions[].destinations.onSuccess` is configured with `Ref` directly', () => {
      return runServerless({
        fixture: 'function',
        configExt: {
          functions: {
            basic: {
              destinations: {
                onSuccess: { Ref: 'SomeSNSArn' },
              },
            },
          },
        },
        command: 'package',
      }).catch((error) => {
        expect(error).to.have.property('code', 'DESTINATIONS_MISSING_ARN');
      });
    });
```

and run the test.
but response is below.

```
  1) lib/plugins/aws/package/compile/functions/index.test.js
       Validation
         should throw error when `functions[].destinations.onSuccess` is configured with `Ref` directly:

      AssertionError: expected ServerlessError: Configuration error at '… { …(5) } to have property 'code' of 'DESTINATIONS_MISSING_ARN', but got 'INVALID_NON_SCHEMA_COMPLIANT_CONFIGUR…'
      + expected - actual

      -INVALID_NON_SCHEMA_COMPLIANT_CONFIGURATION
      +DESTINATIONS_MISSING_ARN
```

I'm not sure what conditions cause `INVALID_NON_SCHEMA_COMPLIANT_CONFIGURATION`... (so I omitted this new test case.)
Could you help me if you have any ideas?

ALOHACREPES345 commented:
I realized that configuration validation causes `INVALID_NON_SCHEMA_COMPLIANT_CONFIGURATION`.
In all test cases, `configValidationMode` is `'error'`, and configValidation error is prior to my validation error.

I found that the default setting of `configValidationMode` is `error` from v4, but current release version is v3 and the default setting of v3 is `warn`. 

I think we should maybe set `warn` in just this test case if we could.

But no other test cases use `warn` setting, so it is a little bit difficult to decide.
How do you think?

pgrzesik commented on lib/plugins/aws/package/compile/functions.js:
```diff
@@ -650,38 +648,78 @@ class AwsCompileFunctions {
 
+  getDesinationsArn(destinationsProperty) {
+    if (typeof destinationsProperty === 'object') {
+      if (!destinationsProperty.arn) {
```
After thinking about it more, I think we can skip these checks and errors as it's syntax that is not recognized by our validation schema. I'll bring this up internally tomorrow and will report back

ALOHACREPES345 replied:
OK. I wait for the response.

pgrzesik replied:
Let's skip these checks and rely only on schema validation 👍 Thanks for patience on this one 🙇

ALOHACREPES345 replied:
OK

pgrzesik reviewed (commented):
Thanks @ALOHACREPES345 - it looks good overall, I have some comments, please let me know what do you think

ALOHACREPES345 commented:
I fixed them.
It would be appreciated if you could review again.

pgrzesik commented on lib/plugins/aws/package/compile/functions.js:
```diff
@@ -650,38 +648,60 @@ class AwsCompileFunctions {
+  }
+
   // Memoized in a constructor
   ensureTargetExecutionPermission(functionAddress) {
```
Let's maybe replace the name `functionAddress` with something more generic? It's a bit confusing at the moment as it's not really a function address in most cases

ALOHACREPES345 replied:
important.
sorry for my lack of consideration.😣

pgrzesik commented on docs/providers/aws/guide/serverless.yml.md:
```diff
@@ -668,10 +668,14 @@ functions:
     # Lambda destination settings
     destinations:
-      # Function name or ARN of target (EventBridge/SQS/SNS topic)
+      # Function name or ARN(or reference) of target (EventBridge/SQS/SNS topic)
```
let's put a space before the opening paren `ARN (or reference)`

pgrzesik commented on docs/providers/aws/guide/serverless.yml.md:
```diff
@@ -668,10 +668,14 @@ functions:
       onSuccess: functionName
-      # Function name or ARN of target (EventBridge/SQS/SNS topic)
-      onFailure: xxx:xxx:target
+      # Function name or ARN(or reference) of target (EventBridge/SQS/SNS topic)
```
same here

pgrzesik reviewed (changes requested):
thanks @ALOHACREPES345 🙇 I have minor suggestions and we should be good to go

ALOHACREPES345 commented:
Thanks again. 🙇‍♂️
I'd appreciate if you can check this when you have time.

pgrzesik reviewed (approved):
Thanks a lot @ALOHACREPES345, amazing job 🙇
